pkg/datastore/postgres/repositories/header_repository.go

```diff
@@ -128,13 +128,21 @@ func (repository HeaderRepository) getHeaderHash(header core.Header) (string, er
 	return hash, err
 }
 
-func (repository HeaderRepository) insertHeader(header core.Header) (int64, error) {
+// Function is public so we can test insert being called for the same header
+// Can happen when concurrent processes are inserting headers
+// Otherwise should not occur since only called in CreateOrUpdateHeader
+func (repository HeaderRepository) InternalInsertHeader(header core.Header) (int64, error) {
 	var headerId int64
-	err := repository.database.QueryRowx(
-		`INSERT INTO public.headers (block_number, hash, block_timestamp, raw, eth_node_id, eth_node_fingerprint) VALUES ($1, $2, $3::NUMERIC, $4, $5, $6) RETURNING id`,
-		header.BlockNumber, header.Hash, header.Timestamp, header.Raw, repository.database.NodeID, repository.database.Node.ID).Scan(&headerId)
+	row := repository.database.QueryRowx(
+		`INSERT INTO public.headers (block_number, hash, block_timestamp, raw, eth_node_id, eth_node_fingerprint)
+		VALUES ($1, $2, $3::NUMERIC, $4, $5, $6) ON CONFLICT DO NOTHING RETURNING id`,
+		header.BlockNumber, header.Hash, header.Timestamp, header.Raw, repository.database.NodeID, repository.database.Node.ID)
+	err := row.Scan(&headerId)
 	if err != nil {
-		log.Error("insertHeader: error inserting header: ", err)
+		if err == sql.ErrNoRows {
+			return 0, ErrValidHeaderExists
```

Comment:
Any corresponding caller error matching, since we're returning an error instead of just logging now?

Reply:
Yep! Header sync will continue on this error https://github.com/vulcanize/vulcanizedb/blob/staging/pkg/history/populate_headers.go#L56